File: app/schemas/template.py
from uuid import UUID

from pydantic import BaseModel


class SchemaResponse(BaseModel):
    template_type: str
    display_name: str
    has_draft: bool = False
    draft_data: dict | None = None
    snapshot_data: dict | None = None
    current_version: int | None = None
    risk_level: str | None = None  # security plan only
    fields: list[dict]


class DraftRequest(BaseModel):
    data: dict


class GenerateRequest(BaseModel):
    data: dict


class GenerateResponse(BaseModel):
    id: UUID
    template_type: str
    version_number: int
    minio_path: str | None = None
    pdf_ready: bool = False
    pdf_preview_url: str | None = None
    created_at: str | None


class VersionItem(BaseModel):
    id: str
    version_number: int
    generated_by: str
    created_at: str | None
    is_current: bool
    pdf_ready: bool = False


class VersionDetail(BaseModel):
    id: str
    version_number: int
    data_snapshot: dict
    template_hash: str
    generated_by: str
    created_at: str | None
    is_current: bool
    pdf_ready: bool = False


class VersionDiff(BaseModel):
    field: str
    old: object
    new: object


class CreateMaterialRequest(BaseModel):
    material_type: str  # "risk_assessment" | "responsibility_letter"


class CreateMaterialResponse(BaseModel):
    material_id: UUID
    name: str
    template_type: str

File: frontend/src/types/template.ts
export interface FieldDef {
  name: string;
  ui_label: string;
  ui_type: "text" | "textarea" | "number" | "date" | "select" | "repeater" | "signature" | "checkbox";
  required?: boolean;
  min?: number;
  max?: number;
  min_items?: number;
  options?: string[];
  condition?: string;
  auto_calc?: string;
  validate?: { pattern: string; message: string };
}

export interface SchemaResponse {
  template_type: string;
  display_name: string;
  has_draft: boolean;
  draft_data: Record<string, unknown> | null;
  snapshot_data: Record<string, unknown> | null;
  current_version: number | null;
  risk_level: string | null;
  fields: FieldDef[];
}

export interface VersionItem {
  id: string;
  version_number: number;
  generated_by: string;
  created_at: string | null;
  is_current: boolean;
  pdf_ready: boolean;
}

export interface VersionDetail extends VersionItem {
  data_snapshot: Record<string, unknown>;
  template_hash: string;
}

export interface VersionDiff {
  field: string;
  old: unknown;
  new: unknown;
}

export interface GenerateResponse {
  id: string;
  template_type: string;
  version_number: number;
  minio_path: string;
  pdf_ready: boolean;
  pdf_preview_url: string | null;
  created_at: string | null;
}
